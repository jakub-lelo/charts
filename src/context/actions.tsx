export const ADD_PLAYER = "ADD_PLAYER";
export const REMOVE_PLAYER = "REMOVE_PLAYER";
export const SUBTRACT_PLAYER_POINTS = "SUBTRACT_PLAYER_POINTS";
export const SET_GAME_STATE = "SET_GAME_STATE";
export const SET_PLAYER_NAME = "SET_PLAYER_NAME";

export interface IPlayer {
    name: string;
    points: number;
}

export enum GameState {
    "notStarted",
    "started",
    "finished",
}

export interface AppAction {
    type: typeof ADD_PLAYER | typeof REMOVE_PLAYER | typeof SET_PLAYER_NAME | typeof SUBTRACT_PLAYER_POINTS | typeof SET_GAME_STATE,
    payload?: any
}

export const addPlayer = (): AppAction => ({
    type: ADD_PLAYER
})

export const removePlayer = (player: number): AppAction => ({
    type: REMOVE_PLAYER,
    payload: player
})

export const subtractPlayerPoints = (points: number): AppAction => ({
    type: SUBTRACT_PLAYER_POINTS,
    payload: points
})

export const setGameState = (state: GameState): AppAction => ({
    type: SET_GAME_STATE,
    payload: state
})

export const setPlayerName = (payload: { name: string, index: number }): AppAction => ({
    type: SET_PLAYER_NAME,
    payload,
})

