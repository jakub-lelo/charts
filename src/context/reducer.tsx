import {AppState} from "./context";
import {AppAction, GameState, IPlayer} from "./actions";

export const initialPlayer = (number: number): IPlayer => (
    {name: `Player ${number}`, points: 501}
)

export const initialState: AppState = {
    players: [initialPlayer(1), initialPlayer(2)],
    currentPlayer: 0,
    gameState: GameState.notStarted
}

export const appReducer = (state: AppState, action: AppAction): AppState => {

    switch (action.type) {
        case "ADD_PLAYER":

            return {...state, players: [...state.players, initialPlayer(state.players.length + 1)]}

        case "REMOVE_PLAYER":

            const newPlayers = [...state.players];
            newPlayers.splice(action.payload, 1);
            return {...state, players: [...newPlayers]}

        case "SET_PLAYER_NAME":

            return {
                ...state, players: state.players.map((player, index) =>
                    action.payload.index === index ?
                        {
                            ...player, name: action.payload.name
                        } : player)
            }

        case "SET_GAME_STATE":

            return {...state, gameState: action.payload}

        case "SUBTRACT_PLAYER_POINTS":

            const player = {...state.players[state.currentPlayer]};

            player.points = player.points - action.payload;

            if (player.points === 0) {
                return {...state, gameState: GameState.finished}

            } else {

                const players = [...state.players];

                if (player.points > 0) {
                    players[state.currentPlayer] = player;
                }

                return {
                    ...state,
                    players,
                    currentPlayer: state.currentPlayer + 1 < state.players.length ? state.currentPlayer + 1 : 0
                }
            }
    }
}