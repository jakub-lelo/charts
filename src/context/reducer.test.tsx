import React from 'react';
import {appReducer, initialState} from "./reducer";
import {addPlayer, GameState, setGameState, setPlayerName, subtractPlayerPoints} from "./actions";

describe('reducer tests', () => {
    it('ADD_PLAYER', () => {
        const action = addPlayer();
        const newState = appReducer(initialState, action);
        expect(newState.players).toHaveLength(3);
        expect(newState).toMatchSnapshot();
    });

    // tbd
    // it('REMOVE_PLAYER', () => {
    //     const action = removePlayer();
    //     const newState = appReducer(initialState, action);
    // });


    it('SET_PLAYER_NAME', () => {
        const action = setPlayerName({name: "Test Player", index: 0});
        const newState = appReducer(initialState, action);
        expect(newState.players[0].name).toMatch("Test Player");
        expect(newState).toMatchSnapshot();
    });

    it('SET_GAME_STATE', () => {
        const action = setGameState(GameState.finished);
        const newState = appReducer(initialState, action);
        expect(newState.gameState).toBe(GameState.finished);
        expect(newState).toMatchSnapshot();
    });

    it('SUBSTRACT_PLAYER_POINTS: Substract 1 point', () => {
        const action = subtractPlayerPoints(1);
        const newState = appReducer(initialState, action);
        expect(newState.players[0].points).toBe(500);
        expect(newState.currentPlayer).toBe(1);
        expect(newState).toMatchSnapshot();
    });

    it('SUBSTRACT_PLAYER_POINTS: Substract too many points', () => {
        const action = subtractPlayerPoints(602);
        const newState = appReducer(initialState, action);
        expect(newState.players[0].points).toBe(501);
        expect(newState.currentPlayer).toBe(1);
    });

    it('SUBSTRACT_PLAYER_POINTS: Change from last to the first player', () => {
        const action = subtractPlayerPoints(1);
        // kopiuje stan, a po przecinku nadbijuje konkretna wartosc
        const modifiedState = {...initialState, currentPlayer: 1};
        const newState = appReducer(modifiedState, action);
        expect(newState.currentPlayer).toBe(0);
    });

    it('SUBSTRACT_PLAYER_POINTS: Game over', () => {
        const action = subtractPlayerPoints(10);
        // kopiuje stan, a po przecinku nadbijuje konkretna wartosc
        const modifiedState = {...initialState};
        modifiedState.players[0].points = 10;
        const newState = appReducer(modifiedState, action);
        expect(newState.gameState).toBe(GameState.finished);
    });
});

