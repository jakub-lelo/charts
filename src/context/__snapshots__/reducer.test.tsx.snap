// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer tests ADD_PLAYER 1`] = `
Object {
  "currentPlayer": 0,
  "gameState": 0,
  "players": Array [
    Object {
      "name": "Player 1",
      "points": 501,
    },
    Object {
      "name": "Player 2",
      "points": 501,
    },
    Object {
      "name": "Player 3",
      "points": 501,
    },
  ],
}
`;

exports[`reducer tests SET_GAME_STATE 1`] = `
Object {
  "currentPlayer": 0,
  "gameState": 2,
  "players": Array [
    Object {
      "name": "Player 1",
      "points": 501,
    },
    Object {
      "name": "Player 2",
      "points": 501,
    },
  ],
}
`;

exports[`reducer tests SET_PLAYER_NAME 1`] = `
Object {
  "currentPlayer": 0,
  "gameState": 0,
  "players": Array [
    Object {
      "name": "Test Player",
      "points": 501,
    },
    Object {
      "name": "Player 2",
      "points": 501,
    },
  ],
}
`;

exports[`reducer tests SUBSTRACT_PLAYER_POINTS: 0 1`] = `
Object {
  "currentPlayer": 1,
  "gameState": 0,
  "players": Array [
    Object {
      "name": "Player 1",
      "points": 500,
    },
    Object {
      "name": "Player 2",
      "points": 501,
    },
  ],
}
`;

exports[`reducer tests SUBSTRACT_PLAYER_POINTS: Substract 1 point 1`] = `
Object {
  "currentPlayer": 1,
  "gameState": 0,
  "players": Array [
    Object {
      "name": "Player 1",
      "points": 500,
    },
    Object {
      "name": "Player 2",
      "points": 501,
    },
  ],
}
`;
